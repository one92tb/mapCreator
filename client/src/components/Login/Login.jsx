import React from "react";
import axios from "axios";
import { connect } from "react-redux";
import {
  BrowserRouter as Router,
  Route,
  Link,
  Redirect,
  withRouter
} from "react-router-dom";
import { createUser } from "../../actions/createUser";
import { loginRequest } from "../../actions/loginRequest";

class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      login: "",
      password: "",
      loginStatus: false,
      userId: "",
      isAuthorized: false
    };
  }

  componentDidUpdate(prevProps, prevState) {
    if (!this.state.isAuthorized && this.props.userId) {
      console.log("oh yea!");
      this.setState({
        userId: this.props.userId,
        isAuthorized: true
      }),
        () => {
          console.log(this.state);
        };
    }
  }

  onChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  onSubmit = e => {
    e.preventDefault();
    const createUser = this.props.createUser;
    const loginRequest = this.props.loginRequest;

    const user = {
      login: this.state.login,
      password: this.state.password
    };

    !this.state.loginStatus ? createUser(user) : loginRequest(user);

    console.log(this.props);
  };

  isLogin = status => {
    this.setState({
      loginStatus: status
    });
  };

  render() {
    console.log(this.state.isAuthorized, "state", this.state);
    if (this.state.isAuthorized) {
      console.log("tak");
      return <Redirect to={{ pathname: "/map" }} />;
    } else {
      console.log("nie");
    }
    return (
      <div>
        <button onClick={() => this.isLogin(false)}>Register</button>
        <button onClick={() => this.isLogin(true)}>Login</button>
        <form>
          <input
            type="text"
            name="login"
            onChange={e => this.onChange(e)}
            value={this.state.user}
          />
          <input
            type="text"
            name="password"
            onChange={e => this.onChange(e)}
            value={this.state.password}
          />
          <button onClick={e => this.onSubmit(e)}>
            {this.state.loginStatus ? "Authorization" : "Create Account"}
          </button>
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = {
  createUser,
  loginRequest
};

const mapStateToProps = state => ({
  isAuthorized: state.account.isAuthorized,
  userId: state.account.userId
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
