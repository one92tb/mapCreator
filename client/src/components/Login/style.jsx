import styled from "styled-components";
import { css } from "styled-components";

const Wrapper = styled.div`
  background: #f2f2f2;
  height: 100%;
  width: 100%;
  padding-top: 100px;
`;

const Inner = styled.div`
  width: 400px;
  border: 1px solid #6c757d;
  margin-left: auto;
  margin-right: auto;
  border-radius: 0.25rem;
`;

const Form = styled.form`
  padding: 20px;
`;

const FormGroup = styled.div`
  position: relative;

  &:not(:last-child) {
    margin-bottom: 1rem;
  }
`;

const ErrorMessage = styled.span`
  color: red;
  font-size: 10px;
  display: block;
  margin-left: 2px;
`;

const Label = styled.label`
  display: block;
  margin-left: 2px;
`;

const Button = css`
  cursor: pointer;
  height: 40px;
  width: 165px;
  padding: 0.375rem 0.75rem;
  border: 1px solid #00b8e6;
  color: #fff;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
`;

const RegisterBtn = styled.button`
  ${Button};
  background-color: ${props => (!props.status ? "#00b8e6" : "#4ddbff")};
`;

const LoginBtn = styled.button`
  ${Button};
  background-color: ${props => (props.status ? "#00b8e6" : "#4ddbff")};
`;

const SubmitBtn = styled.button`
  ${Button};
  background-color: #00b8e6;
  margin-top: 4px;
  width: 100%;
`;

const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  border-bottom: 1px solid #6c757d;
  width: 100%;
  height: 80px;
  padding: 20px;
`;

const Input = styled.input`
  width: 100%;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  padding: 0.375rem 0.75rem;
  display: block;
  border: 1px solid #ced4da;
`;

export {
  Wrapper,
  Inner,
  ButtonWrapper,
  RegisterBtn,
  LoginBtn,
  Form,
  FormGroup,
  Label,
  Input,
  ErrorMessage,
  SubmitBtn
};
