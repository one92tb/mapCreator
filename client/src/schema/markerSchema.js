export const errors = {
  markerNameError: "",
  markerImageFileError: ""
};

export const markerValidationDetails = data => {
  const { markerName, markerImageFile, selectedMarker } = data;

  return [
    {
      condition: markerName.length < 3,
      nameOfErrorProperty: "markerNameError",
      messageError: "Username needs to be atleast 3 characters long"
    },
    {
      condition: !/\.(png)$/i.test(markerImageFile.name) && markerImageFile,
      nameOfErrorProperty: "markerImageFileError",
      messageError: "format must be .png"
    },
    {
      condition: !markerImageFile && !selectedMarker.id,
      nameOfErrorProperty: "markerImageFileError",
      messageError: "Input file cannot be empty"
    }
  ];
};
