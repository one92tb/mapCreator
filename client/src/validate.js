const validate = (errors, registerValidationDetails, data) => {
  let isError = false;

  registerValidationDetails(data).forEach(validate => {
    if (validate.condition) {
      isError = true;
      errors[validate.nameOfErrorProperty] = validate.messageError;
    } else {
      errors[validate.nameOfErrorProperty] = "";
    }
  });

  return {
    isError,
    errors
  };
};

export default validate;
