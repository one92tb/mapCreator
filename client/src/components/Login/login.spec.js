import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { Login } from "./Login";
import validate from "../../validate.js";
import { errors, registerValidationDetails } from "../../schema/registerSchema";

describe("login component", () => {
  let wrapper;

  it("passses all props to Login component", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("onChange", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    const loginInput = wrapper.find("input").first();
    const passwordInput = wrapper.find("input").at(1);

    loginInput.simulate("change", {
      target: {
        name: "login",
        value: "user1"
      }
    });
    passwordInput.simulate("change", {
      target: {
        name: "password",
        value: "asd123"
      }
    });

    expect(wrapper.instance().state.login).toBe("user1");
    expect(wrapper.instance().state.password).toBe("asd123");
  });

  it("onSubmit - create user (success)", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );
    wrapper.setState({
      loginStatus: false
    });
    const submitBtn = wrapper.find("button").at(2);

    wrapper.setState({
      login: "user1",
      password: "asd123AAAAA"
    });

    submitBtn.simulate("click", { preventDefault() {} });

    expect(createUser).toHaveBeenCalledTimes(1);
    expect(resetRegisterError).toHaveBeenCalledTimes(1);
    expect(resetRegisterSuccess).toHaveBeenCalledTimes(1);
    expect(resetLoginError).toHaveBeenCalledTimes(1);
  });

  it("onSubmit - create user (failure - validation)", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );
    wrapper.setState({
      login: "us",
      password: "",
      loginStatus: false
    });

    const submitBtn = wrapper.find("button").at(2);

    submitBtn.simulate("click", { preventDefault() {} });

    expect(wrapper.instance().state.loginError).toBe(
      "The login must be minimum 4 characters or longer"
    );
    expect(wrapper.instance().state.passwordError).toBe(
      "The password must be minimum 8 characters or longer"
    );
    expect(resetRegisterError).toHaveBeenCalledTimes(1);
    expect(resetRegisterSuccess).toHaveBeenCalledTimes(1);
    expect(resetLoginError).toHaveBeenCalledTimes(1);
    expect(createUser).toHaveBeenCalledTimes(0);
  });

  it("onSubmit - create user (user is exist)", () => {
    const createUser = jest.fn(() => {
      wrapper.setProps({
        registerError: {
          response: {
            data: {
              registerError: "This login is exist"
            }
          }
        }
      });
    });
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );
    wrapper.setState({
      login: "user123",
      password: "asd123456G",
      loginStatus: false
    });

    const submitBtn = wrapper.find("button").at(2);

    submitBtn.simulate("click", { preventDefault() {} });

    expect(createUser).toHaveBeenCalledTimes(1);
    expect(
      wrapper.instance().props.registerError.response.data.registerError
    ).toBe("This login is exist");
  });

  it("onSubmit - login request (success)", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    const submitBtn = wrapper.find("button").at(2);

    wrapper.setState({
      login: "user1",
      password: "asd123"
    });

    submitBtn.simulate("click", { preventDefault() {} });

    expect(resetRegisterError).toHaveBeenCalledTimes(1);
    expect(resetRegisterSuccess).toHaveBeenCalledTimes(1);
    expect(resetLoginError).toHaveBeenCalledTimes(1);
    expect(loginRequest).toHaveBeenCalledTimes(1);
  });

  it("onSubmit - login request (failure)", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn(() => {
      wrapper.setProps({
        authError: {
          response: {
            data: {
              errorMessage: "login or password is incorrect"
            }
          }
        }
      });
    });
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    wrapper.setState({
      login: "user1",
      password: "asd123"
    });

    const submitBtn = wrapper.find("button").at(2);
    submitBtn.simulate("click", { preventDefault() {} });

    expect(loginRequest).toHaveBeenCalledTimes(1);
  });

  it("isLogin method - change status from login to register", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    const registerBtn = wrapper.find("button").at(1);
    registerBtn.simulate("click");
    expect(wrapper.state().loginStatus).toBe(false);
  });

  it("isLogin method - change status from register to login", () => {
    const createUser = jest.fn();
    const loginRequest = jest.fn();
    const resetLoginError = jest.fn();
    const resetRegisterError = jest.fn();
    const resetRegisterSuccess = jest.fn();
    const isAuthorized = false;
    const authError = null;
    const registerError = null;
    const registerSuccess = "";

    wrapper = shallow(
      <Login
        createUser={createUser}
        loginRequest={loginRequest}
        resetLoginError={resetLoginError}
        resetRegisterError={resetRegisterError}
        resetRegisterSuccess={resetRegisterSuccess}
        isAuthorized={isAuthorized}
        authError={authError}
        registerError={registerError}
        registerSuccess={registerSuccess}
      />
    );

    wrapper.setState({
      loginStatus: false
    });

    const loginBtn = wrapper.find("button").first();
    loginBtn.simulate("click");
    console.log(loginBtn.props());
    expect(wrapper.state().loginStatus).toBe(true);
  });
});
