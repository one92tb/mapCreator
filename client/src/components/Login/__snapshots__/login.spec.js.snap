// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component passses all props to Login component 1`] = `
<div>
  <div>
    <div>
      <button
        onClick={[Function]}
        status={true}
      >
        Login
      </button>
      <button
        onClick={[Function]}
        status={true}
      >
        Register
      </button>
    </div>
    <form>
      <div>
        <label>
          Login
        </label>
        <input
          autocomplete="login"
          name="login"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <label>
          Password
        </label>
        <input
          autocomplete="current password"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <span />
      </div>
      <div>
        <button
          onClick={[Function]}
        >
          Authorization
        </button>
      </div>
    </form>
  </div>
</div>
`;
